### ISSUE TITLE
How to handle randomness?

### ISSUE THREAD

**Issue by wouterkool** [N-Wouda/PyVRP#163]
Note: this issue is just to have a discussion, I'm fine if we agree the current approach is fine.

Relating to #157, I think it would be helpful if we could explicitly pass around a `rng` object so it is clear which components are deterministic or not, and we can avoid components having random state. This could be a general type where `XorShift12`8 is just one implementation.

We could even consider a 'functional' approach where the `rng` object is actually immutable but can be forked when you need to seed a function call (this is e.g. how [JAX handles randomness](https://jax.readthedocs.io/en/latest/jax-101/05-random-numbers.html)). This improves reproducability and makes it easier to parallelize code without changing the result (as you can split the rng object). I like this idea as it makes randomness easier to reason about (i.e. passing the same immutable `rng` object to a function would give the same result twice) and I think PyVRP can benefit from parallelization (i.e. multiple local searches in parallel), but it may be a bit overkill as well. Just wanted to raise it :)

**Comment by N-Wouda** [comment]
Also relates to #27.

**Comment by N-Wouda** [comment]
One thing that might be possible is to hook into numpy via the [API](https://numpy.org/devdocs/reference/random/c-api.html). Then we wouldn't have to roll anything ourselves, but could just ask the user to do all the usual seed stuff via numpy themselves.

Not sure (a) how to do this, (b) whether to do this, or (c) whether it'd be fast enough for us.

**Comment by N-Wouda** [comment]
I did some more reading and speed would probably be OK. The how bit is basically: get the relevant numpy headers and libraries from where the package's installed, and link against those when compiling. So that should also be possible. I think it's worth exploring, because that'd mean we can get rid of our custom RNG in favour of something much more standard.

Since we're already deeply dependent on numpy anyway, I think it'd be OK to do this.

**Comment by wouterkool** [comment]
We could also try to get all randomness at the Python level and make the C++ api deterministic, which would also make it easier to reason about and test, and easier to customize by the user.
* I think LocalSearch only uses `rng` to shuffle the order of nodes/routes and nodeops/routeops, which we could also set from Python. This would also make the code more flexible and easier to customize (i.e. a user could try to optimize the node or nodeop order).
* For initializing the individual, we could just create the random routes in a Python `make_individual` function (I think I already saw that coming by somewhere?).
* For Individual::SelectiveRouteExchange we could just pass `startA` and `nMovedRoutes` from Python by wrapping the function in Python.
* If we really need a random number generator in C++ (if we would need a lot of random numbers), then we could just pass a seed which we sample from Python and use an internal `rng` object.

**Comment by N-Wouda** [comment]
> I am in favor of using something standard, and `numpy` would be a good option but I'm strongly in favor of avoiding global random state (this would become a mess if we would go parallel) but instead explicitly pass around a [Generator](https://numpy.org/doc/stable/reference/random/generator.html).

That's what I already suggested above.

> Rather than interfacing in C++ with the numpy API, I think it would not be that hard to get all randomness at the Python level and make the C++ api deterministic, which would make it simpler, easier to reason about, easier to test and easier to customize by the user.
> * I think LocalSearch only uses `rng` to shuffle the order of nodes/routes and nodeops/routeops, which we could also set from Python. This would also make the code more flexible and easier to customize (i.e. a user could try to optimize the node or nodeop order).
> * For initializing the individual, we could just create the random routes in a Python `make_individual` function (I think I already saw that coming by somewhere?).
> * For `Individual::SelectiveRouteExchange` we could just pass `startA` and `nMovedRoutes` from Python by wrapping the function in Python.
> * If we really need a random number generator in C++ (if we would need a lot of random numbers), then we could just pass a seed which we sample from Python and use an internal `rng` object.

How would we shuffle the node/route ops? We don't have an easy way to access to those LS internals from Python. Going down this route would break a lot of encapsulation, particularly on the `LocalSearch` object, for IMHO questionable benefit.

**Comment by N-Wouda** [comment]
For `numpy.random`, these are relevant:
- [This SO q/a](https://stackoverflow.com/questions/64139525/numpy-implementing-a-custom-rng)
- [This PR](https://github.com/GalSim-developers/GalSim/pull/1179)
- [This issue?](https://github.com/GalSim-developers/GalSim/issues/1067)
- [The `numpy.random` C API docs](https://numpy.org/doc/stable/reference/random/c-api.html)

**Comment by N-Wouda** [comment]
I have a stash with some trials for this. I think we can make this work with `numpy.random`. Would also be nice because then we get access to a whole gamut of generators/distributions to sample from.

**Comment by N-Wouda** [comment]
I'm closing this as https://github.com/PyVRP/PyVRP/labels/wontfix because it's not really been an issue and I think our efforts are better spent elsewhere.
